File: img_bmp/aritmetica/aritmetica_con_buffer/aritmetica.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

// alineamiento de memoria a 1 byte debido a los distintos dipos de datos
#pragma pack(push, 1)

typedef struct {
    uint16_t type;
    uint32_t size;
    uint16_t reserved1;
    uint16_t reserved2;
    uint32_t offset;
    uint32_t header_size;
    int32_t width;
    int32_t height;
    uint16_t planes;
    uint16_t bitsperpixel;
    uint32_t compression;
    uint32_t image_size;
    int32_t x_resolution;
    int32_t y_resolution;
    uint32_t used_colors;
    uint32_t important_colors;
} BMPHeader;

//alineamiento normal
#pragma pack(pop)

void createBMP(int width, int height) {
    
    // padding = (4 - (width * 3) % 4) % 4;
    uint32_t n_padding_byte = 4 - ((width * 3) % 4);
    if (n_padding_byte == 4){
        n_padding_byte = 0;
    }

    // datos de la cabecera
    BMPHeader header = {
        .type = 0x4D42, // 0x4D42 = MB, se almacena en formato little-endian
        .size = sizeof(BMPHeader) + ((width * 3) + n_padding_byte) * height,
        .reserved1 = 0,
        .reserved2 = 0,
        .offset = sizeof(BMPHeader),
        .header_size = 40,
        .width = width,
        .height = height,
        .planes = 1,
        .bitsperpixel = 24,
        .compression = 0,
        .image_size = ((width * 3) + n_padding_byte) * height,
        .x_resolution = 0,
        .y_resolution = 0,
        .used_colors = 0,
        .important_colors = 0
    };

    // apertura del archivo
    FILE* file = fopen("img_c.bmp", "wb");
    if (!file) {
        printf("Error al abrir el archivo.\n");
        return;
    }

    // escritura de la cabecera en el archivo
    fwrite(&header, sizeof(BMPHeader), 1, file);

    // uso de un buffer inicializado en 0
    size_t data_size = ((width * 3) + n_padding_byte) * height;
    char *data = (char*)calloc(data_size, sizeof(char));
    int index = 0;

    // colores de la imagen
    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            data[index + 2] = (char)((x / (float)width) * 255);
            data[index + 1] = (char)(((height - y) / (float)height) * 255);
            data[index] = (char)((y / (float)height) * 255);
            index += 3;

        }
        index += n_padding_byte;
        
    }

    fwrite(data, sizeof(char), data_size, file);
    free(data);
    fclose(file);
}

int main() {
    int width = 8000;
    int height = 8000;

    createBMP(width, height);
    return 0;
}

File: img_bmp/aritmetica/aritmetica_con_buffer/aritmetica.java
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class aritmetica {
    
    public static void createBMP(int width, int height) {
        
        // padding = (4 - (width * 3) % 4) % 4;
        int n_padding_byte = 4 - ((width * 3) % 4);
        if (n_padding_byte == 4) {
            n_padding_byte = 0;
        }

        int size = 54 + ((width * 3) + n_padding_byte) * height;
        int image_size = ((width * 3) + n_padding_byte) * height;

        // alineamirnto usando un buffer
        ByteBuffer buffer = ByteBuffer.allocate(54).order(ByteOrder.LITTLE_ENDIAN);

        buffer.putShort((short) 0x4D42); // type
        buffer.putInt(size);             // size
        buffer.putShort((short) 0);      // reserved1
        buffer.putShort((short) 0);      // reserved2
        buffer.putInt(54);                // offset
        buffer.putInt(40);                // header_size
        buffer.putInt(width);             // width
        buffer.putInt(height);            // height
        buffer.putShort((short) 1);       // planes
        buffer.putShort((short) 24);      // bitsperpixel
        buffer.putInt(0);                 // compression
        buffer.putInt(image_size);        // image_size
        buffer.putInt(0);                 // x_resolution
        buffer.putInt(0);                 // y_resolution
        buffer.putInt(0);                 // used_colors
        buffer.putInt(0);                 // important_colors
        
        // apertura del archivo
        try (FileOutputStream file = new FileOutputStream("img_java.bmp")) {
            
            // escritura de cabecera
            file.write(buffer.array());
            
            // uso de un bufffer, por defecto se inicializa a 0
            int data_size = ((width * 3) + n_padding_byte) * height;
            byte[] data = new byte[data_size];
            int index = 0;

            // colores del imagen
            for (int y = 0; y < height; y++) {
                for (int x = 0; x < width; x++) {
                    data[index + 2] = (byte) ((x / (double) width) * 255);
                    data[index + 1] = (byte) (((height - y) / (double) height) * 255);
                    data[index] = (byte) ((y / (double) height) * 255);

                    index += 3;
                }
                index += n_padding_byte;
            }
            
            file.write(data);

        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    public static void main(String[] args){
        int width = 8000;
        int height = 8000;
        createBMP(width, height);
    }
}
